cpp/src/arrow/dataset/scanner.h

```diff
@@ -150,34 +153,55 @@ class ARROW_DS_EXPORT SimpleScanner : public Scanner {
   std::shared_ptr<ScanContext> context_;
 };
 
+/// \brief ScannerBuilder is a factory class to construct a Scanner. It is used
+/// to pass information, notably a potential filter expression and a subset of
+/// columns to materialize.
 class ARROW_DS_EXPORT ScannerBuilder {
  public:
   ScannerBuilder(std::shared_ptr<Dataset> dataset,
                  std::shared_ptr<ScanContext> scan_context);
 
-  /// \brief Set
-  ScannerBuilder* Project(const std::vector<std::string>& columns);
-
-  ScannerBuilder* Filter(std::shared_ptr<Expression> filter);
-  ScannerBuilder* Filter(const Expression& filter);
-
-  ScannerBuilder* FilterEvaluator(std::shared_ptr<ExpressionEvaluator> evaluator);
+  /// \brief Set the subset of columns to materialize.
+  ///
+  /// This columns be passed down to DataSources and corresponding DataFragments.
+  /// The goal is to avoid copying/deserializing columns that will be
+  /// ignored/dropped further down the compute chain.
+  ///
+  /// \param[in] columns list of columns to project. Order and duplicates will
+  ///            be preserved.
+  ///
+  /// \return Failure if any column names does not exists in the dataset's
```

Comment:
```suggestion
  /// \return Failure if any column name does not exists in the dataset's
```

Please address partition keys as well